---
title: "NGS analysis for gene regulation and epigenomics"
author: "Jacques Serizay, Cyril Matthey-Doret"
date: "January 11, 2020"
output:
  ioslides_presentation: default
  slidy_presentation: default
---

# Chromosome conformation capture

Day 1

## Capturing chromosome conformation

* 3C methods capture thre 3D structure of the genome
* Originally relied on qPCR, Hi-C is the NGS variant

## Interpreting chromosome contact maps

* Contact frequency reflects spatial organization
* Various patterns correspond to 3D structures



## Derivative methods

3C-based methods:

* MicroC: DNAse instead of restriction enzyme
* Promoter capture: Hi-C with capture probes on promoters

Others:

* GAM: Cryoslicing of nuclei followed by sequencing
* SPRITE: Serial dilution and barcoding


## From reads to contact maps

Many "end-to-end" pipelines available:

* **nf-core/hic**
* Hi-C pro
* FanC
* hicstuff
* Juicer
* ...

Important to understand individual steps (biases, custom analyses, ...)


## Hi-C processing: Read alignment

* Standard short read mapping
* Separate single-end alignment for forward and reverse
* Longer reads -> more chimeric reads
  + Iterative alignment to solve the problem

## Hi-C processing: Merging BAMs

* Merge forward and reverse reads by name
* Filter out ambiguous alignment (low mapQ)
* 1 pair of reads = 1 contact
* The [pairs](https://github.com/4dn-dcic/pairix/blob/master/pairs_format_specification.md) format is commonly used

## Hi-C processing: Fragment attribution

* Assign contacts to discrete segments (bins)
* Usually regular intervals of e.g. 10kb

## Hi-C processing: The Matrix

* Sum contact events for each combination of bins
* Most combinations have 0 contacts: "sparse data"
* Storing NxN matrix with mostly 0 is impractical
  + Use sparse format instead

## Hi-C processing: Additional filters

Hi-C specific filters to remove uninformative events

> - Self-religating fragments
> - Undigested fragments
> - Duplicates filter

## Hi-C processing: Balancing

Hi-C is susceptible to many biases affecting local coverage such as:

* GC content
* Chromatin accessibility
* Restriction site density

Matrix balancing is a normalization to reduce the impact of those biases

## Hi-C processing: Balancing

Assumption: All bins (regions) have the same contact probability.

Divide each pixel by row x col iteratively.

# Exercises

* Generate your own contact map from the reads, normalize and visualize it
