---
title: "Processing ChIP-seq data with Nextflow"
output: 
    rmarkdown::html_document: 
        theme: simplex
        highlight: tango
        preserve_yaml: true
        code_folding: hide
        df_print: tibble
        toc: true
        toc_float: true
---

## 1. Running the nf-core/chipseq pipeline

### Documentation 

Go to the Nextflow Github page and find the ChIP-seq processing workflow.  
Have a look at each processing step and understand what each step does. 

### Test run 

> Try to run a test workflow: 

```{sh eval  = FALSE}
nextflow run nf-core/chipseq -profile test,singularity
```

> Check out the `results/` directory.  

* What is the required input for the pipeline? Can you find it in the online documentation?
* What are the outputs of the pipeline? 

### Run pipeline on data 

#### Prepare sample datasheet 

> Use the samples information file located in `shared/` to build the required input datasheet. 

The following gist can be used to reformat the sample information into an input datasheet. 

```{r}
samples <- read.table('shared/ChIP-seq_samples.txt', header = TRUE, sep = '\t')
samples
samples$ctl <- paste0("control_", samples$Assay)[match(samples$ctl, samples$Run)]
samples$Transcription_factor[samples$Transcription_factor == 'Neg'] = paste0("control_", samples$Assay)[samples$Transcription_factor == 'Neg']
tab <- data.frame(
    group = samples$Transcription_factor,
    replicate = samples$Replicate,
    fastq_1 = samples$reads_1,
    fastq_2 = samples$reads_2,
    antibody = samples$Transcription_factor,
    control = samples$ctl
)
tab[is.na(tab)] = ''
tab$control[tab$control == ''] <- 'control_ChIP-Exo'
tab$control[grepl('control', tab$group)] <- ''
tab$antibody[grepl('control', tab$group)] <- ''
write.csv(tab, 'ChIP-seq_design.csv', quote = FALSE, row.names = FALSE)
tab
```

#### Launch pipeline 

> Execute the nf-core/chipseq pipeline for a single TF. 

```{sh eval  = FALSE}
NXF_OPTS='-Xms1g -Xmx4g'
nextflow run nf-core/chipseq \
    -profile singularity \
    -resume \
    --input shared/ChIP-seq_design.csv \
    --genome 'R64-1-1' \
    --max_cpus 4 \
    --max_memory '28.GB' \
    --save_trimmed true \
    --narrow_peak true \
    --skip_plot_fingerprint \
    --skip_spp
```

## 2. Inspecting output of Nextflow workflow

### Efficiency of adaptor trimming 

Sequenced libraries can sometimes be contaminated with adapters, which are initially ligated to 
pulled-down DNA fragments to enable fragment hybridization on Illumina sequencing chips. 
`trim_galore` is a tool designed to remove suched contaminating adapter sequences. It scans `.fastq` files 
and trim each read if an adapter sequence is found in it. 

> Have a look at the `Cse4` fastq QC reports before and after adapter trimming with `trim_galore`.  
> How efficient does `trim_galore` seem to be?

### Other contaminating sequences

> Inspect Hsf1 ChIP-seq library (replicate 2). Has `trim_galore` removed all the adapter sequences?  
> What about other contaminating sequences? Can you speculate on the origin of such sequences?

<details><summary>Clue</summary>
<p>
Is there any difference in the experimental approaches used to ChIP Hsf1 and Cse4?
</p>
</details>

### ChIP-seq signal

When performing ChIP-seq, one usually tries to profile genome-wide transcription factor binding patterns. 
One of the most important outputs of this pipeline is the `.bigwig` files. These files 

> How is `.bigWig` file structure? Is there a more "human-readable" version of `.bigWig` files? 

<details><summary>Clue</summary>
<p>
Check out documentation provided by UCSC:  
See [bigWig](http://genome.ucsc.edu/goldenPath/help/bigWig.html) and [wig](http://genome.ucsc.edu/goldenPath/help/wiggle.html) documentation.
</p>
</details>

> How can you parse a `.bigWig` file in R? 

<details><summary>Clue</summary>
<p>
Check out rrtacklayer package from Bioconductor:
See [bigWig](http://genome.ucsc.edu/goldenPath/help/bigWig.html) and [wig](http://genome.ucsc.edu/goldenPath/help/wiggle.html) documentation.
</p>
</details>

### ChIP-seq background

You can also load the tracks directly in IGV and visually inspect them.  

> Do you see differences in the background levels?  
> Do you see 

### Conclusions 

> Can you draw general conclusions on whether a specific ChIP-seq assay may be better than the others?

