---
title: "3. Day 03"
day: "Day 03"
---

## Presentations for {{< param "day" >}}

- ChIP-seq assay: Probing for epigenetic marks **\[25min\]**
- Mapping peaks in ChIP-seq datasets **\[25min\]**
- R/Bioconductor: integrated and comprehensive analysis of genomic data **\[25min\]**
- Nextflow: an efficient workflow manager for reproducible research in biological projects **\[25min\]**

## Exercises for {{< param "day" >}}

- Processing ChIP-seq datasets: from A to Z with Nextflow workflow manager **\[55min\]**: [[Rmd]](/{{<myPackageUrl>}}Exercices/day03/processing_ChIPseq.html)

    Data for this part of the course was collected from multiple research papers, in which ChIP-seq (or variants of) 
    experiments have been conducted in Yeast, in normal conditions (no specific treatment).  

    * 
    * 
    * 
    * 

- Motif enrichment analysis with MEME workflow **\[55min\]**

    > Motifs enriched in TF-occupied loci can be identified using motif enrichment tools such as the [MEME suite](http://meme-suite.org/).

- Integrating ChIP-seq results **\[55min\]**

    > Depending on the biological question, TF ChIP-seq results can be integrated to test hypotheses and uncover biological insights. In this case, we will focus on the interactions between the different transcription factors ChIP-ed and analyzed in this course. 