---
title: "Processing ChIP-seq data with Nextflow"
output: 
    rmarkdown::html_document: 
        theme: simplex
        highlight: tango
        preserve_yaml: true
        code_folding: hide
        df_print: tibble
        toc: true
        toc_float: true
---

## 1. Running the nf-core/chipseq pipeline

### Documentation 

Go to the Nextflow Github page and find the ChIP-seq processing workflow.  
Have a look at each processing step and understand what each step does. 

### Test run 

> Try to run a test workflow: 

```{sh eval  = FALSE}
nextflow run nf-core/chipseq -profile test,singularity
```

> Check out the `results/` directory.  

* What is the required input for the pipeline? Can you find it in the online documentation?
* What are the outputs of the pipeline? 

### Run pipeline on data 

#### Prepare sample datasheet 

> Use the samples information file located in `shared/` to build the required input datasheet. 

```{r}
samples <- read.table('shared/ChIP-seq_samples.txt', header = TRUE, sep = '\t')
samples
samples$ctl <- paste0("control_", samples$Assay)[match(samples$ctl, samples$Run)]
samples$Transcription_factor[samples$Transcription_factor == 'Neg'] = paste0("control_", samples$Assay)[samples$Transcription_factor == 'Neg']
tab <- data.frame(
    group = samples$Transcription_factor,
    replicate = samples$Replicate,
    fastq_1 = samples$reads_1,
    fastq_2 = samples$reads_2,
    antibody = samples$Transcription_factor,
    control = samples$ctl
)
tab[is.na(tab)] = ''
tab$control[tab$control == ''] <- 'control_ChIP-Exo'
write.csv(tab, 'ChIP-seq_design.csv', quote = FALSE, row.names = FALSE)
tab
```

#### Launch pipeline 

> Execute the nf-core/chipseq pipeline for a single TF. 

```{sh eval  = FALSE}
NXF_OPTS='-Xms1g -Xmx4g'
nextflow run nf-core/chipseq \
    -profile singularity \
    --input shared/ChIP-seq_design.csv \
    --genome 'R64-1-1' \
    --max_cpus 4 \
    --max_memory '28.GB' \
    --fingerprint_bins 100000 \
    --save_trimmed true \
    --narrow_peak true \
    --skip_plot_fingerprint \
    --skip_spp
```
