---
title: "Ex. 03-1: Processing ChIP-seq data with Nextflow"
output: 
    rmdformats::readthedown: 
        highlight: tango
        preserve_yaml: true
        df_print: tibble
        toc_depth: 4
        css: ../../../custom.css
    rmdformats::downcute: 
        preserve_yaml: true
    rmarkdown::html_document: 
        theme: simplex
        highlight: tango
        preserve_yaml: true
        code_folding: hide
        df_print: tibble
        toc: true
        toc_float: true
---

## 1. Running the nf-core/chipseq pipeline

### Documentation 

Go to the Nextflow Github page and find the ChIP-seq processing workflow.  
Have a look at each processing step and understand what each step does. 

### Test run 

> Try to run a test workflow: 

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{sh eval  = FALSE}
nextflow run nf-core/chipseq -profile test,singularity
```
</p></details><br>

> Check out the `results/` directory.  

> What is the required input for the pipeline? Can you find it in the online documentation?  

> What are the outputs of the pipeline? 

### Run pipeline on data 

#### Prepare sample datasheet 

> Use the samples information file located in `shared/` to build the required input datasheet. 

The following gist can be used to reformat the sample information into an input datasheet. 

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r}
samples <- read.table('shared/ChIP-seq_samples.txt', header = TRUE, sep = '\t')
samples
samples$ctl <- paste0("control_", samples$Assay)[match(samples$ctl, samples$Run)]
samples$Transcription_factor[samples$Transcription_factor == 'Neg'] = paste0("control_", samples$Assay)[samples$Transcription_factor == 'Neg']
tab <- data.frame(
    group = samples$Transcription_factor,
    replicate = samples$Replicate,
    fastq_1 = samples$reads_1,
    fastq_2 = samples$reads_2,
    antibody = samples$Transcription_factor,
    control = samples$ctl
)
tab[is.na(tab)] = ''
tab$control[tab$control == ''] <- 'control_ChIP-Exo'
tab$control[grepl('control', tab$group)] <- ''
tab$antibody[grepl('control', tab$group)] <- ''
write.csv(tab, 'ChIP-seq_design.csv', quote = FALSE, row.names = FALSE)
tab
```
</p></details><br>

#### Launch pipeline (for only one TF, for a matter of time)

> Execute the nf-core/chipseq pipeline for a single TF. 

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{sh eval  = FALSE}
NXF_OPTS='-Xms1g -Xmx4g'
grep -P "^group|^Reb1|^control_ChIP-Exo" shared/ChIP-seq_design.csv > shared/ChIP-seq_subset_design.csv
nextflow run nf-core/chipseq \
    -profile singularity \
    --input shared/ChIP-seq_subset_design.csv \
    --genome 'R64-1-1' \
    --max_cpus '1' \
    --max_memory '8.GB' \
    --save_trimmed false \
    --narrow_peak true \
    --skip_plot_fingerprint \
    --skip_spp
```
</p></details><br>

> Execute the nf-core/chipseq pipeline for a single TF (for a single-end CHIP-seq) 

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{sh eval  = FALSE}
NXF_OPTS='-Xms1g -Xmx4g'
nextflow run nf-core/chipseq \
    -profile singularity \
    -resume \
    --input shared/ChIP-seq_design_2.csv \
    --genome 'R64-1-1' \
    --single_end \
    --max_cpus '5' \
    --max_memory '28.GB' \
    --save_trimmed false \
    --narrow_peak true \
    --skip_plot_fingerprint \
    --skip_spp
```
</p></details><br>

## 2. Inspecting output of Nextflow workflow

### Efficiency of adaptor trimming 

Sequenced libraries can sometimes be contaminated with adapters, which are initially ligated to 
pulled-down DNA fragments to enable fragment hybridization on Illumina sequencing chips. 
`trim_galore` is a tool designed to remove suched contaminating adapter sequences. It scans `.fastq` files 
and trim each read if an adapter sequence is found in it. 

> Have a look at the `Cse4` fastq QC reports before and after adapter trimming with `trim_galore`.  

> How efficient does `trim_galore` seem to be?

### Other contaminating sequences

> Inspect Hsf1 ChIP-seq library (replicate 2). Has `trim_galore` removed all the adapter sequences?  

> What about other contaminating sequences? Can you speculate on the origin of such sequences?

<details><summary>Clue</summary>
<p>
Is there any difference in the experimental approaches used to ChIP Hsf1 and Cse4?
</p>
</details>

### ChIP-seq coverage

#### bigWig files

When performing ChIP-seq, one usually tries to profile genome-wide transcription factor binding patterns.  
One of the most important outputs of this pipeline is the `.bigwig` files. These files are essentially 
a long array of numerical values representing the (normalized) genome-wide coverage of the ChIP-seq 
library. 

> How exactly is `.bigWig` file structure? Is there a more "human-readable" version of `.bigWig` files? 

<details><summary>Clue</summary>
<p>
Check out documentation provided by UCSC:  
See [bigWig](http://genome.ucsc.edu/goldenPath/help/bigWig.html) and [wig](http://genome.ucsc.edu/goldenPath/help/wiggle.html) documentation.
</p>
</details>

> How comparable are `.bigwig` files from one to another? 

<details><summary>Clue</summary>
<p>
Have a look at the content of the folder `results/bwa/mergedLibrary/bigwig/scale/`. 
</p>
</details>

> How can you parse a `.bigWig` file in R? 

<details><summary>Clue</summary>
<p>
Check out rrtacklayer package from Bioconductor:
See [bigWig](http://genome.ucsc.edu/goldenPath/help/bigWig.html) and [wig](http://genome.ucsc.edu/goldenPath/help/wiggle.html) documentation.
</p>
</details>

#### Enrichment over gene features

ChIP-seq coverage over specific gene features (e.g. TSS, TTS or gene bodies) is typically 
plotted as average signal or as a heatmap. 

> Have a look at the metaplots of different TFs. Can you observe differences in the enrichment profiles
for different factors? If so, comment on the differences. 

### ChIP-seq background

You can also load the tracks directly in IGV and visually inspect them.  

> Do you see differences in the background levels?  

> Visually, can you detect specific peak patterns?

<details><summary>Clue</summary>
<p>
Load all the tracks from Reja et al., 2015. Which type of genes have peaks? And which combinations of factors can you identify?  
Load the tracks from Cook et al., 2012. What do you observe?  
Now load Cse4 track. Is the coverage different? Where is it enriched at? Where is Cse4 supposed to be enriched in nuclei?  
</p>
</details>

### Conclusions 

> Can you make preliminary conclusions on which factors are co-localizing on the chromatin? 

> How can you decide whether a given ChIP-seq library was successful or not? 

> Can you draw general conclusions on whether a specific ChIP-seq assay may be better than the others?

