---
title: "Extra resources"
---

Analyzing NGS data can be a complex process, especially with the rise of multi-omics approaches.  
Here is a list of resources we thought would be useful for people interested in going deeper in the analysis of NGS to study gene regulation and epigenomics. 

## Bioinformatics 

### R 

- ***The*** excellent R guide for beginners, by Emmanuel Paradis: [PDF](https://cran.r-project.org/doc/contrib/Paradis-rdebuts_en.pdf)
- ***The*** 150+ pages comprehensive book to learn everything about Bioconductor. This ebook has been published by [Kasper D. Hansen](https://www.ashansenlab.com/) and is freely available under the [CC BY-NC-SA 4.0 license](https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en_US): [PDF](/{{<myPackageUrl>}}docs/bioconductor.pdf)

## Scientific readings

- [2014 Nat. Methods paper from Bioconductor core team describing important object classes](https://www.nature.com/articles/nmeth.3252)
- [To see from how far Bioc comes from\.\.\.](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2004-5-10-r80)
- [More research on osmotic stress reponse](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4405355/)

## Hi-C tools

* A github repository keeping an updated and curated list of Hi-C tools: https://github.com/mdozmorov/HiC_tools
